server/src/main/java/io/crate/execution/jobs/SharedShardContexts.java

```diff
@@ -102,9 +105,14 @@ public SharedShardContext getOrCreateContext(ShardId shardId) throws IndexNotFou
         SharedShardContext sharedShardContext = allocatedShards.get(shardId);
         if (sharedShardContext == null) {
             IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
-            sharedShardContext = new SharedShardContext(indexService, shardId, readerId, wrapSearcher);
-            allocatedShards.put(shardId, sharedShardContext);
-            readerId++;
+            synchronized (this) {
+                sharedShardContext = allocatedShards.get(shardId);
```

Comment:
Other `allocatedShards` accesses within the file afaik also need to be synchronized to ensure this works.

Reply:
Thank you, It does look like `createContext()` needs to be synchronized but if there is a scenario where `createContext()` and `getOrCreateContext()` race for `allocatedShards`, shouldn't we fix it by not calling `createContext()`? Considering `createContext()` as a lightweight version of `getOrCreateContext()`.

Reply:
Oh I guess we need synchronization for `allocatedShards.put()` calls.